### PR TITLE
Atomic-registry : Add/remove rolebinding for users

### PR DESCRIPTION
**Pull request by sub-mod**
depends on https://github.com/cockpit-project/cockpit/pull/4022
TODO write testcases

### CODE REVIEW

File: pkg/kubernetes/scripts/policy.js [hunk 1]
```diff
@@ -190,7 +190,69 @@
...
+                    if(roleBindings[i].name === role) {
+                        roleBinding = roleBindings[i].roleBinding;
+                        break;
+                    }
+                }
+                removeFromArray(roleArray(roleBinding, "subjects"), subject);
+                removeFromArray(roleArrayKind(roleBinding, subject.kind), subject.name);
+                return methods.patch(defaultPolicybinding, patchData);  
+            }
+            function addRoleToPolicyBinding(policyBinding, project, role, subject) {
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/policy.js]: This is actually about 'addSubjectToPolicyBinding' and should be named that way. The role may or may not exist already.

File: pkg/kubernetes/scripts/policy.js [hunk 2]
```diff
@@ -190,7 +190,69 @@
 
                 return methods.create([binding], namespace);
             }
+            function removeRoleFromPolicyBinding(policyBinding, project, role, subject) {
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/policy.js]: This is actually about 'removeSubjectToPolicyBinding' and should be named that way. The role may or may not exist already.

File: pkg/kubernetes/scripts/policy.js (hunk 1 above)
Comment (stefwalter) [inline pkg/kubernetes/scripts/policy.js]: Isn't this function a duplicate of addToRole below. I don't think it's necessary if so.

File: pkg/kubernetes/scripts/policy.js (hunk 2 above)
Comment (stefwalter) [inline pkg/kubernetes/scripts/policy.js]: This function seems like a duplicate of removeFromRole() below. I don't think it's necessary if so.

File: pkg/kubernetes/scripts/projects.js [hunk 3]
```diff
@@ -94,12 +96,95 @@
         }
     ])
 
+    .factory("projectUtil", [
+        function() {
+            function toName(object) {
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: Lets just make this a top level function in this file. If it's not used by anything outside this file.

**Comment by stefwalter** [comment]
This needs a rebase

File: pkg/kubernetes/scripts/projects.js [hunk 4]
```diff
@@ -94,12 +96,95 @@
...
+                return displayRole;
+            }
+            function getOcRolesList() {
+                var ocRoles = [];
+                angular.forEach(registryRoles, function(r) {
+                    ocRoles.push(r.ocRole);
+                });
+                return ocRoles;
+            }
+            function getAllRoles(member, project) {
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: This should have a unit test.

File: pkg/kubernetes/scripts/projects.js [hunk 5]
```diff
@@ -94,12 +96,95 @@
...
+                var displayRole;
+                for (i = registryRoles.length - 1; i >= 0; i--) {
+                    if(registryRoles[i].ocRole === ocRole) {
+                        displayRole = registryRoles[i].displayRole;
+                        break;
+                    }
+                }
+                return displayRole;
+            }
+            function getOcRolesList() {
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: This should have a unit test.

File: pkg/kubernetes/scripts/projects.js [hunk 6]
```diff
@@ -94,12 +96,95 @@
...
-        function(select, loader, policy) {
+        function(putil, select, loader, policy) {
+            var registryRoles = [{ ocRole: "registry-admin", displayRole :"Admin"},
+                { ocRole:"registry-editor", displayRole :"Push" },
+                { ocRole:"registry-viewer", displayRole :"Pull" }];
 
+            function getRegistryRolesMap() {
+                return registryRoles;
+            }
+            function getDisplayRole(ocRole) {
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: This should have a unit test.

File: pkg/kubernetes/scripts/projects.js [hunk 7]
```diff
@@ -94,12 +96,95 @@
...
+                var roleBinds = subjectRoleBindings(member, projectName);
+                var roleBind, meta, ret = [];
+                angular.forEach(roleBinds, function(roleBind) {
+                    meta = roleBind.metadata || { };
+                    if (meta.name)
+                        ret.push(meta.name);
+                });
+                return ret;
+            }
+            function getRegistryRoles(member, project) {
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: This should have a unit test.

File: pkg/kubernetes/scripts/projects.js [hunk 8]
```diff
@@ -94,12 +96,95 @@
...
+                angular.forEach(roleBinds, function(roleBind) {
+                    meta = roleBind.metadata || { };
+                    if (meta.name && ocRegistryRoles.indexOf(meta.name)!== -1) {
+                        roles.push(getDisplayRole(meta.name));
+                    }
+                        
+                });
+                return roles;
+            }
+            function isRegistryRole(member, role, project) {
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: This should have a unit test.

File: pkg/kubernetes/scripts/projects.js [hunk 9]
```diff
@@ -94,12 +96,95 @@
...
+                var oc_roles = getRegistryRoles(member, project);
+                if(oc_roles.indexOf(role) !== -1) {
+                    return true;
+                }
+                return false;
+            }
+            function isRoles(member, namespace) {
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: This should have a unit test.

**Comment by sub-mod** [comment]
unittest and integration test passed

+ journalctl -o cat SYSLOG_IDENTIFIER=kernel 2>&1 | grep 'type=14.*audit' || true
ok 1 testProjectPolicy (__main__.TestRegistry)

# TESTS PASSED [44s on mysystem]

**Pull request closed without merging**